src/Cmts.ml

```diff
@@ -309,10 +314,39 @@ let partition_after_prev_or_before_next t ~prev cmts ~next =
       List.group cmtl ~break:(fun (_, l1) (_, l2) ->
           not (is_adjacent t l1 l2) )
     with
-    | [cmtl] when is_adjacent t (snd (List.last_exn cmtl)) next ->
+    | [cmtl]
+      when is_adjacent t (snd (List.last_exn cmtl)) next
+           && not (is_adjacent t prev loc) ->
         (CmtSet.empty, cmts)
     | after :: befores ->
-        (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
+        let loc_prev = prev.Warnings.loc_start in
+        let loc_prev_end = prev.Warnings.loc_end in
+        let loc_cmt = (snd (List.hd_exn after)).Warnings.loc_start in
+        let loc_cmt_end = (snd (List.hd_exn after)).Warnings.loc_end in
+        let loc_next = next.Warnings.loc_start in
+        if
+          is_adjacent t prev loc
+          && not (is_adjacent t (snd (List.last_exn cmtl)) next)
+        then (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
+        else if
+          Position.column loc_prev = Position.column loc_cmt
+          && Position.column loc_next <> Position.column loc_cmt
+        then (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
+        else if
+          loc_prev_end.pos_lnum = loc_cmt.pos_lnum
+          && loc_cmt_end.pos_lnum <> loc_next.pos_lnum
+        then (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
+        else if
+          loc_cmt.pos_lnum - loc_prev_end.pos_lnum
+          < loc_next.pos_lnum - loc_cmt_end.pos_lnum
+        then (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
+        else if
+          is_adjacent t prev loc
+          && is_adjacent t (snd (List.last_exn cmtl)) next
+          && loc_prev_end.pos_lnum = loc_cmt.pos_lnum
+          && loc_cmt_end.pos_lnum = loc_next.pos_lnum
+        then (CmtSet.of_list after, CmtSet.of_list (List.concat befores))
```

Comment:
you have  5 branches returning `(CmtSet.of_list after, CmtSet.of_list (List.concat befores))`, any way you can simply this a bit ? or at least have comments describing cases handle by each branch.